woocommerce/woocommerce-ios
# Coupons: Resolve filters and search queries working combined in product selector

Closes: #9098 

## Description

This PR fixes the current Coupons filters implementation by wrapping the current `FilterProductListViewModel.Filters` struct into a `CurrentValueSubject` type, so filters and search queries work properly when combined.

Why? Subscribing to the publisher of a `@Published` variable gets the updated result when it changes, however, the notification is sent before the variable itself is updated (the latest value in `.sink` is in `willSet`, not in `didSet`), so we cannot use this update to trigger other methods that rely on the property, as we're passing the old value.

## Changes
* Changes the `FilterProductListViewModel.Filters` from using a `@Published` property wrapper to use `CurrentValueSubject` instead.
* Removes `.dropFirst()` from the search term publisher. This is necessary so we can combine the filter publisher notifications without having to rely on changes in the search terms happening first.
* Unit Test changes:
  * Inject mock storage on some tests that require synching and inserting products, as these would be flaky and crash when we attempt to mutate from a different context queue (I intend to do the same for all test as an improvement, but perhaps this is more suited for a different PR)
  * Declares `@MainActor` for these same tests, so we're sure they run in the main thread. This is necessary because the action Dispatcher requires it while async-await delegates those to background threads if needed.


## Testing instructions

- Go to Menu > Settings > Experimental toggle > Enable `Coupon Management`
- Go to Menu > Coupons > `+` > `Percentage discount` > All Products
- See that using the "search products" field and/or the "Filter" button now works as expected.

## Screenshots

<img width=450 src="https://user-images.githubusercontent.com/3812076/230409815-60795501-0fc1-467c-a74d-ba47657eb062.gif">


---
